src/api.py

```diff
@@ -0,0 +1,239 @@
+        path = join(upload_folder(), filename)
+        xml.save(path)
+    except Exception as err:
+        sio = StringIO()
+        traceback.print_exc(file=sio)
+        return {
+            'status': conf.ERROR,
+            'code': 'error-uploading-xml',
+            'message': 'an error occured uploading the article xml to be processed',
+            'trace': sio.getvalue(),
+        }, 400 # everything is always the client's fault.
+
+    # generate
+    try:
+        article_json = scraper.main(path, {'override': override})
+        json_filename = filename + '.json'
+        json_path = join(upload_folder(), json_filename)
+        open(json_path, 'w').write(article_json)
+    except Exception as err:
+        sio = StringIO()
+        traceback.print_exc(file=sio)
+        return {
+            'status': conf.ERROR,
+            'code': 'error-scraping-xml',
+            'message': str(err),
+            'trace': sio.getvalue()
+        }, 400
+
+    # validate
+    try:
+        ajson_validate.main(open(json_path, 'r'))
+
+    except jsonschema.ValidationError as err:
+        return {
+            'status': 'invalid',
+            'code': 'invalid-article-json',
+            'message': 'the generated article-json failed validation',
+            'trace': str(err), # todo: any good?
+        }, 400
+    except Exception as err:
+        sio = StringIO()
+        traceback.print_exc(file=sio)
+        return {
+            'status': conf.ERROR,
+            'code': 'error-validating-article-json',
+            'message': 'an error occurred attempting to validate the generated article-json',
+            'trace': sio.getvalue()
+        }, 400
+
+    # send to lax
+    try:
+        msid, version = utils.version_from_path(filename)
+        token = str(uuid.uuid4())
+        args = {
+            # the *most* important parameter. don't modify lax.
+            'dry_run': True,
+
+            # a forced ingest by default
+            'action': 'ingest',
+            'force': True,
+
+            # article details
+            'id': msid,
+            'version': int(version),
+            'article_json': article_json,
+
+            'token': token,
+        }
+        lax_resp = adaptor.call_lax(**args)
+        LOG.info("lax response", extra=lax_resp)
+
+        #api_resp = copy.deepcopy(lax_resp)
+        api_resp = utils.subdict(lax_resp, ['status', 'code', 'message', 'trace'])
+
+        if api_resp['status'] in [adaptor.INVALID, adaptor.ERROR]:
+            # failure
+            return api_resp, 400
+
+        # success
+        # 'code', 'message' and 'trace' are not returned by lax on success, just 'status'
+        api_resp['ajson'] = json.loads(article_json)['article']
+        api_resp['override'] = override
+        return api_resp, 200
+
+    except Exception as err:
+        # lax returned something indecipherable
+        return {
+            'status': conf.ERROR,
+            'code': 'lax-is-borked',
```

Comment:
not sure if client needs to know about `lax`, from their point of view there is some API accepting article XMLs.  Can the `code` be made more generic?